sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinationsHelpers.java

```diff
@@ -265,4 +278,81 @@ public String toString() {
           .toString();
     }
   }
+
+  static <T, DestinationT> DynamicDestinations<T, DestinationT> matchTableDynamicDestinations(
+      DynamicDestinations<T, DestinationT> inner, BigQueryServices bqServices) {
+    return new MatchTableDynamicDestinations<>(inner, bqServices);
+  }
+
+  static class MatchTableDynamicDestinations<T, DestinationT>
+      extends DelegatingDynamicDestinations<T, DestinationT> {
+    private final BigQueryServices bqServices;
+
+    private MatchTableDynamicDestinations(
+        DynamicDestinations<T, DestinationT> inner, BigQueryServices bqServices) {
+      super(inner);
+      this.bqServices = bqServices;
+    }
+
+    private Table getBigQueryTable(TableReference tableReference) {
+      BackOff backoff =
+          BackOffAdapter.toGcpBackOff(
+              FluentBackoff.DEFAULT
+                  .withMaxRetries(3)
+                  .withInitialBackoff(Duration.standardSeconds(1))
+                  .withMaxBackoff(Duration.standardSeconds(2))
+                  .backoff());
+      try {
+        do {
+          try {
+            BigQueryOptions bqOptions = getPipelineOptions().as(BigQueryOptions.class);
+            return bqServices.getDatasetService(bqOptions).getTable(tableReference);
+          } catch (InterruptedException | IOException e) {
+            LOG.info("Failed to get BigQuery table " + tableReference);
+          }
+        } while (nextBackOff(Sleeper.DEFAULT, backoff));
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+      return null;
+    }
+
+    /** Identical to {@link BackOffUtils#next} but without checked IOException. */
+    private static boolean nextBackOff(Sleeper sleeper, BackOff backoff)
+        throws InterruptedException {
+      try {
+        return BackOffUtils.next(sleeper, backoff);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    /** Returns a {@link TableDestination} object for the destination. May not return null. */
+    @Override
+    public TableDestination getTable(DestinationT destination) {
+      TableDestination wrappedDestination = super.getTable(destination);
+      Table existingTable = getBigQueryTable(wrappedDestination.getTableReference());
+
+      if (existingTable == null) {
+        return wrappedDestination;
+      } else {
+        return new TableDestination(
+            wrappedDestination.getTableSpec(),
+            existingTable.getDescription(),
+            existingTable.getTimePartitioning());
```

Comment:
I believe this may also need the table's clustering information? But perhaps DynamicDestinations does not support it yet?